src/Microsoft.OData.Core/ODataBatchWriter.cs

```diff
@@ -422,473 +388,307 @@ void IODataBatchOperationListener.BatchOperationContentStreamDisposed()
+        /// <param name="contentId">The Content-ID value to write in ChangeSet head.</param>
+        protected abstract void VerifyCanCreateOperationRequestMessage(bool synchronousCall, string method, Uri uri,
+            string contentId);
 
         /// <summary>
-        /// Creates an <see cref="ODataBatchOperationResponseMessage"/> for writing an operation of a batch response - implementation of the actual functionality.
+        /// Sets a new writer state; verifies that the transition from the current state into new state is valid.
         /// </summary>
-        /// <param name="contentId">The Content-ID value to write in ChangeSet head.</param>
-        /// <returns>The message that can be used to write the response operation.</returns>
-        private ODataBatchOperationResponseMessage CreateOperationResponseMessageImplementation(string contentId)
-        {
-            this.WritePendingMessageData(true);
-
-            // In responses we don't need to use our batch URL resolver, since there are no cross referencing URLs
-            // so use the URL resolver from the batch message instead.
-            this.CurrentOperationResponseMessage = ODataBatchOperationResponseMessage.CreateWriteMessage(
-                this.rawOutputContext.OutputStream,
-                /*operationListener*/ this,
-                this.payloadUriConverter.BatchMessagePayloadUriConverter,
-                this.container);
-            this.SetState(BatchWriterState.OperationCreated);
-
-            Debug.Assert(this.currentOperationContentId == null, "The Content-ID header is only supported in request messages.");
+        /// <param name="newState">The writer state to transition into.</param>
+        protected abstract void SetState(BatchWriterState newState);
 
-            // write the operation's start boundary string
-            this.WriteStartBoundaryForOperation();
+        /// <summary>
+        /// Validates that the batch writer is ready to process a new write request.
+        /// </summary>
+        protected abstract void ValidateWriterReady();
 
-            // write the headers and request separator line
-            ODataBatchWriterUtils.WriteResponsePreamble(this.rawOutputContext.TextWriter, changeSetBoundary != null, contentId);
+        /// <summary>
+        /// Write any pending headers for the current operation message (if any).
+        /// </summary>
+        /// <param name="reportMessageCompleted">
+        /// A flag to control whether after writing the pending data we report writing the message to be completed or not.
+        /// </param>
+        protected abstract void WritePendingMessageData(bool reportMessageCompleted);
 
-            return this.CurrentOperationResponseMessage;
-        }
+        /// <summary>
+        /// Writes the start boundary for an operation. This is either the batch or the changeset boundary.
+        /// </summary>
+        protected abstract void WriteStartBoundaryForOperation();

```

Comment:
This should be moved to the derived class; it's never called from the base class.

Reply:
cleaned up. Thanks.

---
In reply to: [141141776](https://github.com/OData/odata.net/pull/960#discussion_r141141776) [](ancestors = 141141776)